controller/monitor/snapshot_monitor.go

```diff
@@ -0,0 +1,588 @@
+	task, ok := arg.(snapshotCheckTask)
+	if !ok {
+		return fmt.Errorf("failed to assert value: %v", arg)
+	}
+
+	if !m.addToInProgressSnapshotCheckTasks(task.snapshotName) {
+		return nil
+	}
+	defer m.deleteFromInProgressSnapshotCheckTasks(task.snapshotName)
+
+	engine, err := m.ds.GetVolumeCurrentEngine(task.volumeName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get engine for volume %v", task.volumeName)
+	}
+
+	if err := m.checkVolumeNotInMigration(engine.Spec.VolumeName); err != nil {
+		return err
+	}
+
+	engineCliClient, err := m.getEngineBinaryClient(engine.Spec.VolumeName)
+	if err != nil {
+		return err
+	}
+
+	engineClientProxy, err := engineapi.GetCompatibleClient(engine, engineCliClient, m.ds, m.logger, m.proxyConnCounter)
+	if err != nil {
+		return err
+	}
+	defer engineClientProxy.Close()
+
+	m.Lock()
+	m.SnapshotCheckState = longhorn.SnapshotCheckStateInProgress
+	m.Unlock()
+
+	defer func() {
+		m.Lock()
+		defer m.Unlock()
+		m.SnapshotCheckState = longhorn.SnapshotCheckStateIdle
+	}()
+
+	err = m.requestSnapshotHashing(engine, engineClientProxy, task.snapshotName, task.changeEvent)
+	if err != nil {
+		return err
+	}
+
+	return m.waitForSnapshotHashingComplete(engine, engineClientProxy, task.snapshotName)
+}
+
+func (m *SnapshotMonitor) requestSnapshotHashing(engine *longhorn.Engine, engineClientProxy engineapi.EngineClientProxy,
+	snapshotName string, changeEvent bool) error {
+	rehash := false
+	if !changeEvent {
+		dataIntegrity, err := m.getSnapshotDataIntegrity(engine.Spec.VolumeName)
+		if err != nil {
+			return err
+		}
+		if dataIntegrity == longhorn.SnapshotDataIntegrityEnabled {
+			rehash = true
+		}
+	}
+	return engineClientProxy.SnapshotHash(engine, snapshotName, rehash)
+}
+
+func (m *SnapshotMonitor) waitForSnapshotHashingComplete(engine *longhorn.Engine, engineClientProxy engineapi.EngineClientProxy,
+	snapshotName string) error {
+	opts := []retry.Option{
+		retry.Context(m.ctx),
+		retry.Attempts(snapshotHashSyncStatusAttempts),
+		retry.DelayType(retry.FixedDelay),
+		retry.LastErrorOnly(true),
+		retry.Delay(snapshotHashSyncStatusPeriod * time.Second),
+		retry.RetryIf(func(err error) bool {
+			if err == nil {
+				return false
+			}
+			return strings.Contains(err.Error(), string(engineapi.ProcessStateInProgress))
+		}),
+	}
+
+	if err := retry.Do(func() (err error) {
+		return m.syncHashStatusFromEngineReplicas(engine, engineClientProxy, snapshotName)
+	}, opts...); err != nil {
+		return errors.Wrapf(err, "failed to sync hash status for snapshot %v since %v", snapshotName, err)
+	}
+
+	return nil
+}
+
+func (m *SnapshotMonitor) checkVolumeNotInMigration(volumeName string) error {
+	v, err := m.ds.GetVolume(volumeName)
+	if err != nil {
+		return err
+	}
+	if v.Spec.MigrationNodeID != "" {
+		return fmt.Errorf("cannot operate during migration")
+	}
+	return nil
+}
+
+func (m *SnapshotMonitor) getEngineBinaryClient(volumeName string) (client *engineapi.EngineBinary, err error) {
```

Comment:
NIT: Do we have an existing function for this? It seems that I have seen this somewhere else.

Reply:
Yes, in `manager/engine.go`, `controller/volume_controller.go` and `controller/backup_controller.go`.
If I would like to refactor it. Which package is better for `getEngineBinaryClient`?

Reply:
I think it's `engineapi`

Reply:
Addressed. Thank you.